Rfc9293 WIP

ebrasca opened the pull request.

ebrasca committed:
tcp: Remove redundant :ack-p t

:ack-p t is the default.

ebrasca committed:
tcp: Update the protocol specification link to rfc9293

ebrasca committed:
tcp: Do nothing to finish segments when in state :closed or :listen

ebrasca committed:
tcp: Fix errata

ebrasca committed:
tcp: Refactor acceptable-segment-p

ebrasca committed:
tcp: Add tcp4-send-ack

ebrasca committed:
tcp: Add challenge-ack

ebrasca committed:
tcp: Check the sequence numbers before accepting RST in :syn-sent

ebrasca committed:
tcp: Send RST when package is from old connection in :syn-sent state

ebrasca committed:
tcp: Refactor tcp4-connection-receive :syn-sent state

ebrasca committed:
tcp: ACK non RST incoming unacceptable segments

ebrasca committed:
tcp: Check incomming RST segments

ebrasca committed:
tcp: Don't abort connection when resiving segment in :syn-received state

ebrasca committed:
tcp: Remove connection from listener when getting SYN in state :syn-received

ebrasca committed:
tcp: Challenge any SYN segment when not in :syn-sent state

ebrasca committed:
tcp: Add :time-wait state

ebrasca committed:
tcp: Add =<

ebrasca committed:
tcp: Deal with wrap around sequence numbers correctly

ebrasca committed:
tcp: Small refactor

ebrasca committed:
tcp: Send RST to segments of old connections in :syn-received state

ebrasca committed:
tcp: Handle FIN in :syn-received state

ebrasca committed:
tcp: Chenck ACK sequence number in :last-ack state before ending it

ebrasca committed:
tcp: Hangle FIN in :last-ack state

ebrasca committed:
tcp: Ignore SYN or RST packets without ACK

ebrasca committed:
tcp: Allow sending data in half open connection

ebrasca committed:
tcp: Add rfc5961 mitigation

ebrasca committed:
tcp: Send ACK to segments that acknowledges something not yet sent

ebrasca committed:
tcp: Hangle FIN in :closing state

ebrasca committed:
tcp: Hangle FIN in :close-wait state

ebrasca committed:
tcp: Allow to resive data and control in the same segment

ebrasca committed:
tcp: Refactor tcp4-connection-receive :fin-wait-1 state

ebrasca committed:
tcp: Update window size

ebrasca committed:
arp: Use correct function to get the time

ebrasca committed:
tcp: Report connection-closing in tcp-send when connetion is closing

ebrasca committed:
tcp: Implemented abort close

TCP specification (rfc9293) says to send a reset segment

ebrasca committed:
tcp: Refactor tcp-send

ebrasca committed:
tcp: Add missing :closed case

Make sure the Mezzano compilation does not fail due to missing :closed case

ebrasca committed:
tcp: Implemented time-wait timeout

ebrasca committed:
tcp: Add :closed case to tcp4-connection-receive

ebrasca committed:
tcp: Add close for :syn-sent and :syn-received cases

fitzsim reviewed (commented):
Nothing here.

fitzsim commented on net/tcp.lisp:
```diff
@@ -645,7 +645,7 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
                 (detach-tcp-connection connection)
                 (tcp4-send-packet connection
-                                  (tcp-connection-snd.next connection)
+                                  (tcp-connection-snd.nxt connection)
```
This one I can approve right away because I have the fix on a local branch; I noticed the runtime issue the typo causes while I was testing the SLIP driver.

fitzsim commented on net/tcp.lisp:
```diff
@@ -700,7 +698,7 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
                              (tcp-connection-receive-event connection))
                             t)
-                      (tcp4-send-packet connection ack (+u32 seq 1) nil :ack-p t))
+                      (tcp4-send-packet connection ack (+u32 seq 1) nil))
```
These changes look good.

fitzsim commented on net/tcp.lisp:
```diff
@@ -1,14 +1,8 @@
 ;;;
 ;;; Transmission Control Protocol - Protocol Specification
-;;; https://tools.ietf.org/html/rfc793
+;;; https://datatracker.ietf.org/doc/html/rfc9293
```
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -379,6 +379,7 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
              (when (not *netmangler-force-local-retransmit*)
                (tcp4-send-packet connection iss (+u32 irs 1) nil :syn-p t))))
           ((logtest flags +tcp4-flag-rst+)) ; Do nothing for resets addressed to nobody.
+          ((logtest flags +tcp4-flag-fin+)) ; Do nothing for finish since the SEG.SEQ cannot be validated
```
Do you have a way to test this?

fitzsim commented on net/tcp.lisp:
```diff
@@ -490,16 +490,15 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
-        (and (and (<= rcv.nxt seg.seq) (< seg.seq (+ rcv.nxt rcv.wnd)))
+        (and (<= rcv.nxt seg.seq)
+             (< seg.seq (+ rcv.nxt rcv.wnd))
              (or (eql seg.len 0)
```
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -490,16 +490,15 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
              (eql seg.seq rcv.nxt))
         ;; Arithmetic here is not wrapping, so as to avoid wrap-around problems.
-        (and (and (<= rcv.nxt seg.seq) (< seg.seq (+ rcv.nxt rcv.wnd)))
+        (and (<= rcv.nxt seg.seq)
```
OK.

fitzsim commented on net/tcp.lisp:
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -779,6 +767,12 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
         (when errors-escape
           (error c))))))
 
+(defun tcp4-send-ack (connection)
```
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -773,6 +773,12 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
                     (tcp-connection-rcv.nxt connection)
                     nil))
 
+(defun challenge-ack (connection)
```
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -591,19 +591,20 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
-                (setf (tcp-connection-snd.una connection) ack)
-                (when (not *netmangler-force-local-retransmit*)
-                  (tcp4-send-packet connection ack (tcp-connection-rcv.nxt connection) nil))
+                (setf (tcp-connection-state connection) :established
```
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -591,19 +591,20 @@ Set to a value near 2^32 to test SND sequence number wrapping.")
                 (setf (tcp-connection-state connection) :syn-received
                       (tcp-connection-rcv.nxt connection) (+u32 seq 1))
-                (when (not *netmangler-force-local-retransmit*)
+                ;; TODO: Update window
```
OK.

fitzsim commented on net/tcp.lisp:
```diff
@@ -760,8 +760,10 @@ to wrap around logic"
                 (challenge-ack connection))
-               ((logtest flags +tcp4-flag-ack+)
-                (detach-tcp-connection connection))))
+               ((not (logtest flags +tcp4-flag-ack+))) ; Ignore packets without ACK set.
```
Typo in the commit message for this one.  Not commenting on the logic in the change.

fitzsim commented on net/tcp.lisp:
```diff
@@ -763,7 +763,11 @@ to wrap around logic"
                (t
                 (when (eql ack (tcp-connection-snd.nxt connection))
-                  (detach-tcp-connection connection)))))
+                  (detach-tcp-connection connection))
```
Another typo in this commit message.

fitzsim commented on net/tcp.lisp:
```diff
@@ -709,8 +709,8 @@ to wrap around logic"
                            (eql seq (tcp-connection-rcv.nxt connection)))
-                  (setf (tcp-connection-rcv.nxt connection) (+u32 seq 1)
-                        (tcp-connection-state connection) :close-wait)
+                  (setf (tcp-connection-state connection) :close-wait
```
Typo in commit message: "resive"..

fitzsim commented on net/arp.lisp:
```diff
@@ -130,7 +130,7 @@ Returns NIL if there is no entry currently in the cache, this will trigger a loo
 
 (defun arp-expiration ()
-  (let ((time (1+ (get-internal-real-time))))
+  (let ((time (1+ (get-universal-time))))
```
Is it an option to use internal-real-time throughout arp.lisp instead?  Is get-universal-time monotonic, or can it change backwards when setting the system date/time (in which case it should not be used for ARP timeouts).

ebrasca committed:
tcp: Default MSS for IPv4 is 536

ebrasca committed:
tcp: Support the mandatory option set

ebrasca committed:
tcp: Set own MSS to 1460

ebrasca committed:
tcp: Add checks for updating send window

https://datatracker.ietf.org/doc/html/rfc9293#section-3.10.7.4-2.5.2.2.2.3.2.2

ebrasca committed:
tcp: Fix errata

ebrasca committed:
tcp: Handle retransmision and timeout in :syn-received state

ebrasca committed:
tcp: Add missing update-window

ebrasca committed:
tcp: Better docstrings for acceptable-segment-p and acceptable-ack-p

ebrasca committed:
tcp: Cancel retransmit after 3-Way Handshake

Fixes crash when accepting listener connection

ebrasca committed:
tcp: Cancel retransmit when getting FIN in :syn-received

ebrasca committed:
tcp: Update snd.nxt when sending FIN

ebrasca committed:
tcp: Refactor implementation for clarity and configurability

- Reorganized code with section headers for better structure
- Added detailed comments for constants and parameters
- Replaced hardcoded MSS values with *default-max-seg-size* and *max-seg-size*
- Replaced `(when (not ...))` with `(unless ...)` for readability
- Improved documentation of TCP flags and options

ebrasca committed:
tcp: Fix simultaneous close handling in :fin-wait-1 state

ebrasca committed:
tcp: Rename max-seg-size to mss

ebrasca committed:
tcp: Use MTU to calculate MSS

ebrasca committed:
tcp: Rename MSS parameters to clarify sender/receiver roles

ebrasca committed:
tcp: Refactor and add documentation

ebrasca committed:
tcp: Add missing rfc5961 mitigation in state :syn-sent

ebrasca committed:
tcp: Use challenge-ack in places where ACKs are defensive

ebrasca committed:
tcp: Document the purpose of tcp4-send-ack and challenge-ack

ebrasca committed:
tcp: Add link to rfc5961

ebrasca committed:
tcp: Use effective send and receive mss

MUST-16 DONE
SHLD-6 DONE as MTU does not change
MUST-67 DONE

ebrasca committed:
tcp: Fix TCP option reading checks

ebrasca committed:
tcp: Add flow control to send

ebrasca committed:
tcp: Refactor rtt-measurement

ebrasca committed:
tcp: Nothing to retransmit in :time-wait or :fin-wait-2 state
